### PR TITLE
Add implementation for intrinsic `Anint`

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Fixes: https://github.com/lfortran/lfortran/issues/3009

### CODE REVIEW

File: src/libasr/pass/intrinsic_function_registry.h [hunk 1]
```diff
@@ -1693,16 +1699,16 @@ namespace Aint {
     static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
             diag::Diagnostics& diagnostics) {
         ASRUtils::require_impl(x.n_args > 0 && x.n_args < 3,
-            "ASR Verify: Call to aint must have one or two arguments",
+            "Call to aint must have one or two arguments",
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:1702]:
```suggestion
            "Call to anint must have one or two arguments",
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 2]
```diff
@@ -1693,16 +1699,16 @@ namespace Aint {
...
             x.base.base.loc, diagnostics);
         ASR::ttype_t *type = ASRUtils::expr_type(x.m_args[0]);
         ASRUtils::require_impl(ASRUtils::is_real(*type),
-            "ASR Verify: Arguments to aint must be of real type",
+            "Argument to aint must be of real type",
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:1706]:
```suggestion
            "Argument to anint must be of real type",
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 3]
```diff
@@ -1693,16 +1699,16 @@ namespace Aint {
...
             x.base.base.loc, diagnostics);
         if (x.n_args == 2) {
             ASR::ttype_t *type2 = ASRUtils::expr_type(x.m_args[1]);
             ASRUtils::require_impl(ASRUtils::is_integer(*type2),
-                "ASR Verify: Second Argument to aint must be of integer type",
+                "Second Argument to aint must be of integer type",
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:1711]: same comment

File: src/libasr/pass/intrinsic_function_registry.h [hunk 4]
```diff
@@ -1728,7 +1734,7 @@ namespace Aint {
             int kind = -1;
             if (!ASR::is_a<ASR::Integer_t>(*expr_type(args[1])) ||
                     !extract_value(args[1], kind)) {
-                err("`kind` argument of the `aint` function must be an "
+                err("`kind` argument of the `aint` function must be a "
                     "scalar Integer constant", args[1]->base.loc);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:1738]:
```suggestion
                    "scalar real constant", args[1]->base.loc);
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 5]
```diff
@@ -1773,6 +1779,95 @@ namespace Aint {
...
+        if (scope->get_symbol(fn_name)) {
+            ASR::symbol_t *s = scope->get_symbol(fn_name);
+            ASR::Function_t *f = ASR::down_cast<ASR::Function_t>(s);
+            return b.Call(s, new_args, expr_type(f->m_return_var), nullptr);
+        }
+        fill_func_arg("a", arg_types[0]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+
+        ASR::expr_t* adjusted_value;
+        if ( fLtE(args[0], f(0.0, arg_types[0])) ) {
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:1856]: This is not how you do, here you are creating `if` in c++ code and not in ASR, to get it done you have to do: 
```suggestion
       ASR::expr_t* test = make_Compare(make_IntegerCompare_t, args[0], GtE, zero);
```
Refer: `instantiate_Abs` and try to understand difference. To see the ASR you generate with the current code do

```
% lfortran a.f90 --pass=intrinsic_functions --show-asr
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 6]
```diff
@@ -1773,6 +1779,106 @@ namespace Aint {
...
+        *     r = aint(x-0.5)
+        * end if
+        */
+       
+        ASR::expr_t *test;
+        ASR::expr_t* zero = make_ConstantWithType(make_RealConstant_t, 0.0, arg_types[0], loc);
+        test = make_Compare(make_RealCompare_t, args[0], Gt, zero);
+    
+        Vec<ASR::stmt_t *> if_body; if_body.reserve(al, 1);
+        if_body.push_back(al, b.Assignment(result, i2r(r2i64(r32Add(args[0], f(0.5, arg_types[0]))), return_type)));
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:1868]:
```suggestion
        if_body.push_back(al, b.Assignment(result, <Function call to Aint>));
```

Refer: `instantiate_Trailz` to see how to create a function call to other intrinsic.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 7]
```diff
@@ -1773,6 +1779,106 @@ namespace Aint {
...
+        Vec<ASR::stmt_t *> else_body; else_body.reserve(al, 1);
+        else_body.push_back(al, b.Assignment(result, i2r(r2i64(r32Sub(args[0], f(0.5, arg_types[0]))), return_type)));
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:1870]: Same comment as above

File: src/libasr/pass/intrinsic_function_registry.h (hunk 6 above)
Comment (Pranavchiku) [reply src/libasr/pass/intrinsic_function_registry.h:1868]: You have to do something like `ASR::expr_t* func_call_aint = Aint::instantiate_Aint(al, loc, scope, arg_types_aint, return_type, new_args_aint, 0);`
and then 
`if_body.push_back(al, b.Assignment(result, func_call_aint));`

**Comment by certik** [comment]
Awesome, looks good!

**Comment by Pranavchiku** [comment]
Can you test `anint(0)` and see if it aligns with GFortran?

**Review by Pranavchiku — APPROVED** [review]
I left a comment above, rest all looks good to me. Thank you!

**Comment by Pranavchiku** [comment]
Checked for cases near zero, output aligns. This is good to merge.

**Comment by HarshitaKalani** [comment]
Thanks!

**Pull request merged**